### DISCUSSION TITLE
How to ensure my highlight links are applied LAST?

### DISCUSSION THREAD

**Discussion by craigmac** [hrsh7th/nvim-cmp#1185] [Q&A] [answered]
I'm trying to set default linkages in ~/.config/nvim/init.lua for the main groups with:

```
vim.cmd([[
hi! link CmpItemAbbrDefault Pmenu
hi! link CmpItemAbbrDeprecatedDefault Comment
hi! link CmpItemAbbrMatchDefault Pmenu
hi! link CmpItemAbbrMatchFuzzyDefault Pmenu
hi! link CmpItemKindDefault Special
hi! link CmpItemMenuDefault Pmenu
]])
```

But when I check any of the with e.g., `hi CmpItemAbbrDefault` I see it's not set to link to Pmenu. 

But it seems nvim-cmp/plugin/cmp.lua is still overruling it (https://github.com/hrsh7th/nvim-cmp/blob/913eb8599816b0b71fe959693080917d8063b26a/plugin/cmp.lua#L28). Can anyone help me understand this logic, I'm not sure why it needs to be have all these inherit calls etc., shouldn't they just link to default and/or preferred highlighting groups?

**Comment by fitrh** [comment, marked as answer]
https://github.com/hrsh7th/nvim-cmp/blob/2427d06b6508489547cd30b6e86b1c75df363411/plugin/cmp.lua#L28-L40

> Can anyone help me understand this logic, I'm not sure why it needs to be have all these inherit calls etc., shouldn't they just link to default and/or preferred highlighting groups?

Above code is syntactic sugar over nvim autocmd and highlight API, the `inherit`
function means `'CmpItemAbbrDefault'` will use `'Pmenu'` values but override 
`bg` and `dafault` attributes

---

> But it seems nvim-cmp/plugin/cmp.lua is still overruling it 

Seems like it is the intended behavior of `Cmp*Default` groups since the 
`default` argument (`:h hi-default`) is false, meaning even if the group 
already specified, cmp will still override them, I think you should 
instead override groups without `Default` e.g. `CmpItemAbbr`

**Reply by craigmac** [reply]
Hmm I'll try that, but all the non-Cmp*Default ones, like CmpItemAbbr, just link to their Cmp*Default ones, not sure why it does that, thought it might be some legacy thing.

**Reply by fitrh** [reply]
> but all the non-CmpDefault ones, like CmpItemAbbr, just link to their CmpDefault ones

IDK, but it is like a fallback if the colorscheme does not define `Cmp*` highlight groups

**Reply by craigmac** [reply]
Anyway, that seems to have worked, thanks!
